python/dgl/heterograph.py

```diff
@@ -6065,6 +6065,88 @@ def local_scope(self):
         self._node_frames = old_nframes
         self._edge_frames = old_eframes
 
+    def new_formats(self, allowed_formats=None, create_formats=None):
+        r"""Get a cloned graph with the specified allowed sparse format(s)
+        and create specific sparse matrices allowed for the graph or query
+        for the usage status of sparse formats.
+
+        The API copies both the graph structure and the features.
+
+        If both ``allowed_formats`` and ``create_formats`` are None, the
+        API returns a dict recording the usage status of sparse formats.
+        In any other case, a DGLGraph is returned. The format(s) of the
+        resulting cloned graph is depends on the ``allowed_formats``,
+        ``create_formats`` and the format(s) of the original graph:
+
+        * If ``create_formats`` isn't None, the cloned graph's sparse
+          formats will be consistent with ``create_formats``.
+        * Otherwise:
+          * If the original graph's format is a subset of ``allowed_formats``,
+            the cloned graph's sparse formats will remain consistent with the
+            original.
+          * If a format in the original graph is also in ``allowed_formats``,
+            any formats not in ``allowed_formats`` will be removed in the cloned
+            graph.
+          * If the original graph has no formats in ``allowed_formats``, a new
+            format will be made for the cloned graph. This will be in the order
+            of ``'coo'``, ``'csr'``, ``'csc'``, regardless of the
+            ``allowed_formats`` order.
```

Comment:
L6081 - 6093 appear too complex for me. Can we simplify the behaviors?

Reply:
That's correct. I believe this aspect requires further consideration. While drafting the document, I also felt this part was overly complex. Perhaps a new API would make the entire design clearer. However, this needs more discussion. @frozenbugs

Reply:
It will be more confusion for user if we just say "a subset of the allowed formats will be created", how can I know which allowed format will be created.

What about: always create all allowed format, if create_formats == None?
if create_formats is a subset of allowed_formats, then create the create_formats
Otherwise, raise an error.

Reply:
It actually doesn't matter. Many ops create allowed sparse formats on demand. None of these ops mention the creation of particular sparse formats in their docs and most users never complain about them. If a sparse format is allowed, then it may be created. That's it. Easy.